### ISSUE TITLE
[FEATURE REQUEST] Export utils, contexts, and constructors

### ISSUE THREAD

**Issue by alansouzati** [styled-components/styled-components#2418]
I understand that the react hooks migration will take sometime. It will most likely be part of the next [major version](https://github.com/styled-components/styled-components/issues/1934) of styled-components.

At the same time, I'm facing [performance issues](https://github.com/styled-components/styled-components/issues/2215) with styled-components. This issue is very pressing, and I'm confident that moving to react hooks will get rid of the problem.

So I decided to take a stab at creating my own version of [StyledComponent](https://github.com/styled-components/styled-components/blob/master/packages/styled-components/src/models/StyledComponent.js) that replaces `*Consumer` with `useContext(...)`. The basic idea is to do something like:

```
import styled from "styled-components-no-context";

const Box = styled.div``; // no consumer will be added because we are using useContext
```

Initially I was very confident that this solution would work, until I've looked into `node_modules/styled-components/dist`. I was really hoping the `esm` distribution would maintain the module exports so that I could just replace `../utils/determineTheme` with `styled-components/lib/utils/determineTheme`.

Instead, you are controlling what gets exported or not. The StyleSheet module is inside a `__DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS` variable.

I'm a very superstitious guy, so I don't want to use that variable. 😆 

Jokes apart, there would be quite a few things I would need to manually copy over: all utils, the constructors, and contexts. You don't expose `StyleSheetContext` so there is no way for me to do `useContext(StyleSheetContext)`.

I'm a big fan of styled-components. At the same time, I don't want to sit and wait for a v5, not even talking about other breaking changes that will come that will prevent me to migrate immediately once it is out.

I decided to take a look at emotion and see how they structure their dist content. I really liked the architecture there where everything lives in its own package, for example:

```
import { createContext, createElement, forwardRef, Component, Fragment } from 'react';
import createCache from '@emotion/cache';
import { getRegisteredStyles, insertStyles } from '@emotion/utils';
import { serializeStyles } from '@emotion/serialize';
import { StyleSheet } from '@emotion/sheet';
import css from '@emotion/css';
export { default as css } from '@emotion/css';
```

Both [styled-components and emotion](https://www.youtube.com/watch?v=AqFqduEEw7A&t=19s) are suffering from the context issue. I feel it will take some time until you both move away from it. In the meantime, as it stands today, emotion offers a better architecture to allow consumers to migrate earlier.

Please don't get me wrong, I want to keep using styled-components, but I'm afraid I will be out of options if I'm not able to fix the performance issues in my organization.

**Comment by probablyup** [comment]
I get what you're saying, but doing this would mean all of the internals become part of the public API and therefore subject to semver. I can't see us doing it.

**Comment by alansouzati** [comment]
Do you have an ETA when you will migrate to hooks?

**Comment by alansouzati** [comment]
can you at least add the StyleSheetContext inside `DO_NOT_USE_OR_YOU_WILL_BE_HAUNTED_BY_SPOOKY_GHOSTS `?

**Comment by tbranyen** [comment]
@probablyup why would it have to be part of the public API? Couldn't you do what you're already doing and namespace it under something silly like:

```
__use_this_and_youre_not_cool__: {
  /* all the internals* /
}
```

Anyone who uses this would know that its not part of the public API, but you'd also help unblock folks who want to hack on Styled Components in their apps. As of now Styled Components is pretty buggy and slow in a lot of areas and nobody external can help improve it for their apps.

Just some additional thoughts.

**Comment by probablyup** [comment]
@alansouzati There are open PRs for it now actually. I've been slammed at work so haven't had a lot of time to do s-c stuff. If you want to help review, that'd be great.

**Comment by probablyup** [comment]
I mean I object to that spooky ghosts thing even being exposed in the first place, honestly. If we get issues complaining that we changed something, that still counts as the public API changing. Even prefixed like that, I can guarantee if we changed an internal we would get angry issue posts.

**Comment by alansouzati** [comment]
yeah, I understand. Another solution is to keep the file structure as is, instead of creating one single file. That way, your `index.js` is what get exposed and if I want to access things directly I can do the manual path. In fact, you already have other folders (e.g. `test-utils`) inside your package, so doing this:

`import { findAll } from "styled-components";`

returns undefined.

But doing this:

`import { findAll } from "styled-components/test-utils";`

returns this util.

IMHO "angry issue posts" should never be what is blocking a feature to be implemented or not.

You already have [other issues](https://github.com/styled-components/styled-components/issues/135#issuecomment-468525090) that can benefit from allowing different instances of styled components. You can also treat this as an "angry post" (see the number of plus ones and people commenting asking for it)...

So if there is a way to easily access your contexts, utils, and constructors, folks can create their preferred API to create their styled components.

```javascript
import styled from "./my-styled";
const Box = styled("div", { excludeProps: ['color'] })`
  color: ${props => props.color};
`;

const Flex = styled(Box, { excludeProps: ['direction'] })`
  flex-direction: ${props => props.direction};
`;
```

I personally think your utils and contexts should be part of the public API.

**Comment by probablyup** [comment]
> IMHO "angry issue posts" should never be what is blocking a feature to be implemented or not.

When you're developing something on volunteer time and not getting paid, it's a factor.

**Comment by alansouzati** [comment]
I do understand the challenge with open-source and the fact that we do that out of love. I feel the same pain in Grommet, and other open source libs I maintain. So, I HEAR YOU THERE.

But I don't understand how this relates to "angry posts" 😄
